File: 2021/21-dirac-dice/solution.hs
import Prelude
import Data.List

rolls = [1..100] ++ rolls

play (_:b:_:_:e:_:xs) p1 p2
  | (snd p1) >= 1000 || (snd p2) >= 1000 = [(p1, p2)]
  | otherwise = (p1, p2):play xs (step p1 (3 * b)) (step p2 (3 * e))

step :: (Int, Int) -> Int -> (Int, Int)
step (p, s) x = let x' = x `mod` 10
                    p' = p + x'
                    p'' = if p' > 10 then p' `mod` 10 else p'
                in (p'', s + p'')

part1 =
  let game = tail $ play rolls (10, 0) (4, 0)
      (s1, s2):(s1', s2'):_ = map (\(p1, p2) -> (snd p1, snd p2)) $ reverse game
      l = 6 * length game
      (rc, loser) = if s1 >= 1000 then (l - 3, s2') else (l, s1)
  in rc * loser

quantomRolls = map (\g -> (length g, head g)) $ group $ sort $ allRolls
  where
    allRolls = do
      d1 <- [1, 2, 3]
      d1' <- [1, 2, 3]
      d1'' <- [1, 2, 3]
      pure (d1 + d1' + d1'')

main = do
  print part1
  print $ quantomRolls

File: 2021/21-dirac-dice/solution.js
const rolls = [[1,3],[3,4],[6,5],[7,6],[6,7],[3,8],[1,9]]

const key = (a, b, c, d) => (b * 10000) + (c * 1000) + (a * 100) + d
const play = (p1, s1, p2, s2, mem) => {
  const k = key(p1, s1, p2, s2)
  if (s2 >= 21) return mem[k] = [0, 1]
  if (mem[k]) return mem[k]

  let sw1 = 0
  let sw2 = 0
  for (const [l, r] of rolls) {
    const roll_p = p1 + (r % 10)
    const new_position1 = roll_p > 10 ? roll_p % 10 : roll_p

    const [w2, w1] = play(p2, s2, new_position1, s1 + new_position1, mem)
    sw1 += w1 * l
    sw2 += w2 * l
  }

  return mem[k] = [sw1, sw2]
}

console.log(play(10, 0, 4, 0, {}).sort((a, b) => b - a))
